aten/src/ATen/StorageImpl.h

```diff
@@ -108,43 +88,111 @@ struct AT_API StorageImpl : public Retainable {
   at::DataPtr& data_ptr() {
     return data_ptr_;
   };
-  void set_data_ptr(at::DataPtr&& data_ptr) {
-    data_ptr_ = std::move(data_ptr);
+  const at::DataPtr& data_ptr() const {
+    return data_ptr_;
+  };
+  // Returns the previous data_ptr
+  at::DataPtr set_data_ptr(at::DataPtr&& data_ptr) {
+    std::swap(data_ptr_, data_ptr);
+    return std::move(data_ptr);
   };
   void* data() {
     return data_ptr_.get();
   };
   const void* data() const {
     return data_ptr_.get();
   };
+  at::DeviceType device_type() const {
+    return data_ptr_.device().type();
+  }
   at::Allocator* allocator() {
     return allocator_;
   };
-  at::ScalarType& scalar_type() {
+  at::ScalarType scalar_type() const {
     return scalar_type_;
   };
   const at::Allocator* allocator() const {
     return allocator_;
   };
-  int getDevice() const {
-    return data_ptr_.device().index();
+  // You generally shouldn't use this method, but it is occasionally
+  // useful if you want to override how a tensor will be reallocated,
+  // after it was already allocated (and its initial allocator was
+  // set)
+  void set_allocator(at::Allocator* allocator) {
+    allocator_ = allocator;
+  }
+  Device device() const {
+    return data_ptr_.device();
   }
   void set_resizable(bool resizable) {
     resizable_ = resizable;
   }
 
+  // You should only call these functions if you have a raw StorageImpl*
+  // pointer; if you have intrusive_ptr<StorageImpl> this will be
+  // handled automatically.
+  //
+  // TODO: Eliminate as many uses of these functions as humanly possible
+  void _raw_incref() {
+    auto ptr = c10::intrusive_ptr<StorageImpl>::reclaim(this);
+    auto ptr_copy = ptr;
+    ptr_copy.release();
+    ptr.release();
+  }
+  void _raw_decref() {
+    // Let it die
+    c10::intrusive_ptr<StorageImpl>::reclaim(this);
+    // NB: You still "have" a pointer, but it's now invalid.
+    // If you want more safety, used the actual c10::intrusive_ptr class
+  }
+  StorageImpl* _raw_make_weak() {
+    // NB: this is a strong reference
+    auto ptr = c10::intrusive_ptr<StorageImpl>::reclaim(this);
+    c10::weak_intrusive_ptr<StorageImpl> wptr(ptr);
+    ptr.release();
+    return wptr.release();
+  }
+  void _raw_weak_incweakref() {
```

Comment:
I think one weak is enough :).

Reply:
It's actually an important distinction.

When I have a `StorageImpl*` there are two types of raw pointer it could be. It could be a *strong* pointer, which is only permissible when `refcount > 0`, or it could be a *weak* pointer, which is permissible for any value of refcount, even when `refcount == 0`. In the old, `Retainable` implementation, we didn't really distinguish between these two cases, which means you could try to increment the (strong) refcount, even when you only had a weak pointer. That would be invalid if `refcount == 0`; once you lose all strong references, you are never allowed to get a strong reference to the object again. But nothing checked for this.

In the new `c10::intrusive_ptr` implementation, the implementations of `intrusive_ptr::reclaim` and `weak_intrusive_ptr::reclaim` implementations have much more stringent asserts: `intrusive_ptr` asserts that there MUST have been a strong pointer which was released; similarly, `weak_intrusive_ptr` asserts that there MUST have been a weak pointer that was released. So, if you for example, call `_raw_weak_incweakref` on a strong pointer, and the object has no weak pointers on it, the assert will conclude that this is an invalid use of `weak_intrusive_ptr::reclaim`, because your `weakcount == 1` (and that weak reference is implied by the strong reference), so you couldn't have possibly release'd a weak pointer that you are now reclaiming, you must have done something wrong.

So, `_raw_weak_incweakref` is ONLY valid if you call it on a "weak" `StorageImpl*` pointer. That's what the first weak means. And it increments the weak reference count; thus the second weak.

Reply:
Can there be a `_raw_incweakref`?  Or `_raw_weak_incref` / `raw_weak_decref`?

Reply:
I guess my assumption was, as you mentioned, there are two regimes of refcounts: weak and strong.  Anyone using the raw interface would know this so say, "raw_weak_incref" would have a clear meaning.  But I can see the argument the other way as well, so do what you think is best!